tools/envoy_headersplit/headersplit.py

```diff
@@ -0,0 +1,303 @@
+
+    """
+  impl_cursors = []
+  for i in cursor.walk_preorder():
+    if i.location.file is None:
+      continue
+    if i.location.file.name != cursor.displayname:
+      continue
+    if i.kind == CursorKind.NAMESPACE:
+      continue
+    if i.semantic_parent is not None and i.semantic_parent.kind == CursorKind.CLASS_DECL:
+      impl_cursors.append(i)
+  return impl_cursors
+
+
+def extract_definition(cursor: Cursor, classnames: List[str]) -> Tuple[str, str, List[str]]:
+  """
+    extracts class definition source code pointed by the cursor parameter.
+    and find dependent mock classes by naming look up.
+
+    Args:
+        cursor: libclang cursor pointing to the target mock class definition.
+        classnames: all mock class names defined in the definition header that needs to be
+            divided, used to parse class dependencies.
+    Returns:
+        class_name: a string representing the mock class name.
+        class_defn: a string contatins the whole class definition body.
+        deps: a set of string contatins all dependent classes for the return class.
+
+    Note:
+        It can not detect and resolve forward declaration and cyclic dependency. Need to address
+        manually.
+    """
+  filename = cursor.location.file.name
+  with open(filename, 'r') as source_file:
+    contents = source_file.read()
+  class_name = cursor.spelling
+  class_defn = contents[cursor.extent.start.offset:cursor.extent.end.offset] + ";"
+  # need to know enclosed semantic parents (namespaces)
+  # to generate corresponding definitions
+  parent_cursor = cursor.semantic_parent
+  while parent_cursor.kind == CursorKind.NAMESPACE:
+    if parent_cursor.spelling == "":
+      break
+    class_defn = "namespace {} {{\n".format(parent_cursor.spelling) + class_defn + "\n}\n"
+    parent_cursor = parent_cursor.semantic_parent
+
+  # resolve dependency
+  # by simple naming look up
+
+  deps = set()
+  for classname in classnames:
+    if classname in class_defn and classname != class_name:
+      deps.add(classname)
+
+  return class_name, class_defn, deps
+
+
+def extract_implementation(cursor: Cursor) -> Tuple[str, int]:
+  """
+    extracts class methods implementation source code pointed by the cursor parameter.
+    and find dependent mock classes by naming look up.
+
+    Args:
+        cursor: libclang cursor pointing to the target mock class definition.
+
+    Returns:
+        class_name: a string representing the mock class name.
+        implline: the first line of the corresponding impl code
+
+    Note:
+        this function return line number only. Because in certain case libclang will fail
+        in parsing the method body and return an empty function body instead. So we choose
+        not to parse the function body, get the start line and the end line instead.
+    """
+  class_name = cursor.semantic_parent.spelling
+  return class_name, cursor.extent.start.line - 1
+
+
+def main(args):
```

Comment:
There's a lot of code here - would it be possible to move some of it to helper functions, and add unit tests for those helpers?

Reply:
I moved some of it to helper functions, unit tests for them is WIP